bee-ternary/src/lib.rs

```diff
@@ -0,0 +1,875 @@
+    /// have. Providing a slice that is not valid for this encoding is undefined behaviour.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if `num_trits` is more than can be represented with the slice in
+    /// the given encoding.
+    ///
+    /// # Safety
+    ///
+    /// This function must only be called with an [`i8`] slice that is valid for this trit encoding
+    /// given the specified `num_trits` length. Right now, this validity is not well-defined and so
+    /// it is suggested that only [`i8`] slices created from existing trit slices or trit buffers
+    /// be used. Calling this function with an invalid [`i8`] slice is undefined behaviour.
+    pub unsafe fn from_raw_unchecked_mut(raw: &mut [i8], num_trits: usize) -> &mut Self {
+        debug_assert!(
+            raw.iter().copied().all(T::is_valid),
+            "Invalid i8 slice used to create trit slice"
+        );
+        &mut *(T::from_raw_unchecked_mut(raw, num_trits) as *mut _ as *mut _)
+    }
+
+    /// Interpret an (`std::i8`) slice as a trit slice with the given encoding, checking to ensure
+    /// that the slice is valid in the given encoding. The `num_trits` parameter is used to specify
+    /// the exact length, in trits, that the slice should be taken to have.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if `num_trits` is more than can be represented with the slice in
+    /// the given encoding.
+    pub fn try_from_raw(raw: &[i8], num_trits: usize) -> Result<&Self, Error> {
+        if raw.iter().copied().all(T::is_valid) {
+            Ok(unsafe { Self::from_raw_unchecked(raw, num_trits) })
+        } else {
+            Err(Error::InvalidRepr)
+        }
+    }
+
+    /// Interpret a mutable (`std::i8`) slice as a mutable trit slice with the given encoding,
+    /// checking to ensure that the slice is valid in the given encoding. The `num_trits` parameter
+    /// is used to specify the exact length, in trits, that the slice should be taken to have.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if `num_trits` is more than can be represented with the slice in
+    /// the given encoding.
+    pub fn try_from_raw_mut(raw: &mut [i8], num_trits: usize) -> Result<&mut Self, Error> {
+        if raw.iter().copied().all(T::is_valid) {
+            Ok(unsafe { Self::from_raw_unchecked_mut(raw, num_trits) })
+        } else {
+            Err(Error::InvalidRepr)
+        }
+    }
+
+    /// Returns `true` if the trit slice is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns the number of trits in this trit slice.
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Interpret this slice as an (`std::i8`) slice.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the slice is not byte-aligned
+    // TODO: Evaluate whether this API makes sense to be `unsafe`
```

Comment:
Legacy comment ?

Reply:
Ah, yes. Turns out that I did evaluate it.